### PR TITLE
Metrics customization guide with .NET 9 Advice API details

### LINKED ISSUE #5855: Update documentation for .NET 9 Advice API
_Originally posted by @cijothomas in https://github.com/open-telemetry/opentelemetry-dotnet/issues/5854#issuecomment-2368882470_

### PR DESCRIPTION
**Pull request by Annosha**
Closes #5855 Update documentation for .NET 9 Advice API

## Changes

- Added documentation for explicit bucket histogram aggregation using the .NET 9 Advice API.
- Provided example code to demonstrate the usage of explicit bucket boundaries in histograms.
- Included best practices for selecting meaningful bucket boundaries and tips for using the new API.

## Merge requirement checklist

* [x] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (license requirements, nullable enabled, static analysis, etc.)
* [ ] Unit tests added/updated
* [ ] Appropriate `CHANGELOG.md` files updated for non-trivial changes
* [ ] Changes in public API reviewed (if applicable)

### CODE REVIEW

**Comment by Annosha** [comment]
@CodeBlanch Please review the changes and provide necessary feedback.

File: docs/metrics/customizing-the-sdk/README.md [hunk 1]
```diff
@@ -335,6 +335,61 @@ by using Views.
 
 See [Program.cs](./Program.cs) for a complete example.
 
+### Explicit Bucket Histogram Aggregation with .NET 9 Advice API
```
Comment (cijothomas) [inline docs/metrics/customizing-the-sdk/README.md:338]: It’s misleading to call it .net 9 advice api as this api can be used in older .net runtimes

File: docs/metrics/customizing-the-sdk/README.md [hunk 2]
```diff
@@ -335,6 +335,61 @@ by using Views.
...
+histogram.Record(25.0);
+```
+
+#### Best Practices
+
+- **Select Meaningful Bucket Boundaries**: When defining bucket
+ boundaries, consider the typical range of values that your application
+ processes. This helps to ensure that the histogram provides relevant
+ insights.
+- **Monitor and Adjust**: After implementing explicit bucket
```
Comment (cijothomas) [inline docs/metrics/customizing-the-sdk/README.md:387]: Without some concrete steps to monitor, I think it’s best to omit this.

File: docs/metrics/customizing-the-sdk/README.md [hunk 3]
```diff
@@ -335,6 +335,61 @@ by using Views.
...
+distribution of values, as it allows for tailored metrics
+collection that aligns with application requirements.
+
+#### Using Explicit Bucket Boundaries
+
+To use explicit bucket boundaries in your histograms, follow these steps:
+
+1. **Install Required Packages**: Ensure you have the latest version of the
+OpenTelemetry .NET SDK that supports the Advice API.
+2. **Define Histogram with Explicit Buckets**: When creating a histogram metric,
```
Comment (cijothomas) [inline docs/metrics/customizing-the-sdk/README.md:359]: One important information here is to document how view api vs advice interact, in particular the priority when both exist. 
I think advice api gets priority.
also mention that if exponential histogram is specified via views, then that takes priority and buckets from hint are ignored

File: docs/metrics/customizing-the-sdk/README.md (hunk 1 above)
Comment (cijothomas) [inline docs/metrics/customizing-the-sdk/README.md:338]: thus document is about sdk customization, so histogram advice is somewhat misfit here. we can mention it in the views section about custom buckets, but anything more should be elsewhere.

Comment (reyang) [inline docs/metrics/customizing-the-sdk/README.md:338]: I doubt if this doc the right place, this doc is focusing on how to customize the SDK. Advice API is part of the instrumentation.

File: docs/metrics/customizing-the-sdk/README.md (hunk 3 above)
Comment (CodeBlanch) [reply docs/metrics/customizing-the-sdk/README.md:359]: Priority when selecting explicit histogram bounds:

* View API (always wins if specified)
* Advice API
* SDK defaults (used when nothing else is specified)

**Comment by Annosha** [comment]
@cijothomas @reyang @CodeBlanch please provide feedback on latest changes.

File: docs/metrics/customizing-the-sdk/README.md [hunk 4]
```diff
@@ -224,6 +224,52 @@ default boundaries. This requires the use of
...
 ```
 
+### Customizing OpenTelemetry .NET SDK for Metrics
+
+#### Using Explicit Buckets in Histograms
+
+In OpenTelemetry .NET, histograms can be customized using the View API, and the Advice API allows additional flexibility for defining explicit bucket boundaries in instrumentation.
+
+
+##### View API vs. Advice API: Defining Histogram Buckets
```
Comment (reyang) [inline docs/metrics/customizing-the-sdk/README.md:234]: I understand where this is coming from. However, I suggest that we take a different approach, instead of having "vs." in our mind, focus on "how these things work together".

This is what I suggest:

1. The actual explicit buckets will be determined by the SDK based on the following factors:
    * The explicit buckets that the instrumentation owner provided via the Advice API.
    * The explicit buckets provided via View API during SDK initialization.
    * The default buckets provided by the SDK.
2. Here is how the SDK will determine the buckets:
    * ...

**Comment by Annosha** [comment]
@reyang please take a look at the improvements.

File: docs/metrics/customizing-the-sdk/README.md [hunk 5]
```diff
@@ -224,6 +224,51 @@ default boundaries. This requires the use of
         new ExplicitBucketHistogramConfiguration { Boundaries = Array.Empty<double>() })
 ```
 
+#### Determining Explicit Buckets for Histograms
+
+In OpenTelemetry .NET, histograms can be customized using both
+the View API and the Advice API. The actual explicit buckets used by
```
Comment (reyang) [inline docs/metrics/customizing-the-sdk/README.md:230]:
```suggestion
the View API and the Advice API. The actual explicit buckets used by
```

File: docs/metrics/customizing-the-sdk/README.md [hunk 6]
```diff
@@ -224,6 +224,51 @@ default boundaries. This requires the use of
...
+1. **View API**: If custom buckets are defined via the View API, these
+take precedence over any buckets provided by the Advice API or SDK defaults.
+2. **Advice API**: If the View API does not specify custom buckets, the SDK
+uses the explicit buckets provided by the Advice API.
+3. **SDK Defaults**: If neither the View API nor the Advice API specifies
+custom buckets, the SDK applies its default configuration.
+
+Additionally, if an exponential histogram is defined using the View API,
+the explicit bucket boundaries provided by the Advice API will be ignored,
+as exponential histograms take precedence.
```
Comment (reyang) [inline docs/metrics/customizing-the-sdk/README.md:248]: Want to make sure I understand this - if the user leveraged the View API to choose a subset of the attributes, and there are no explicit buckets provided - would the buckets provided by the Hint API take effect or not?

Comment (Annosha) [reply docs/metrics/customizing-the-sdk/README.md:248]: Yes, the buckets provided by the Advice (Hint) API would take effect in this case. Here's what I understand:

1. Exponential Histogram with View API: If the user defines an exponential histogram via the View API, it takes full precedence, and any explicit bucket boundaries provided by the Advice API are ignored.

2. No Explicit Buckets Provided by View API: If the View API is used but does not define explicit bucket boundaries (e.g., it only filters attributes or configures other aspects), then:

- The SDK will fall back to the explicit buckets from the Advice (Hint) API if they are available.

- If no buckets are defined in the Advice API either, then the SDK defaults apply.

@cijothomas Could you please verify if these are correct?

File: docs/metrics/customizing-the-sdk/README.md (hunk 1 above)
Comment (cijothomas) [reply docs/metrics/customizing-the-sdk/README.md:383]: This is unresolved.

Comment (cijothomas) [reply docs/metrics/customizing-the-sdk/README.md:383]: this is unresolved.

**Review by cijothomas — CHANGES_REQUESTED** [review]
Few comments left in the PR. My main concern is below:
Customizing the SDK section is not the right place to talk about Advice API in detail. We should mention it, and list the priority order when Advice and View are both used, but nothing more.

File: docs/metrics/customizing-the-sdk/README.md [hunk 7]
```diff
@@ -473,11 +572,11 @@ recording of `Exemplar`s.
 OpenTelemetry SDK comes with the following `ExemplarFilter`s (defined on
 `ExemplarFilterType`):
 
-* (Default behavior) `AlwaysOff`: Makes no measurements eligible for becoming an
+- (Default behavior) `AlwaysOff`: Makes no measurements eligible for becoming an
```
Comment (cijothomas) [inline docs/metrics/customizing-the-sdk/README.md:575]: please remove all unrelated changes from this PR. We are glad to accept them as a separate PR, so as to keep each PR focused on only one aspect only.

File: docs/metrics/customizing-the-sdk/README.md (hunk 2 above)
Comment (cijothomas) [reply docs/metrics/customizing-the-sdk/README.md:432]: @Annosha A lot of comments are marked resolved, but I don't see how. Please leave a comment about how the comment is addressed before resolving them.

File: docs/metrics/customizing-the-sdk/README.md [hunk 8]
```diff
@@ -224,6 +224,51 @@ default boundaries. This requires the use of
...
+custom buckets, the SDK applies its default configuration.
+
+Additionally, if an exponential histogram is defined using the View API,
+the explicit bucket boundaries provided by the Advice API will be ignored,
+as exponential histograms take precedence.
+
+##### Example: Using Explicit Buckets via the Advice API
+
+Here is an example of how to use the Advice API to specify explicit bucket
+ boundaries for a histogram:
```
Comment (reyang) [inline docs/metrics/customizing-the-sdk/README.md:253]:
```suggestion
boundaries for a histogram:
```

File: docs/metrics/customizing-the-sdk/README.md [hunk 9]
```diff
@@ -224,6 +224,51 @@ default boundaries. This requires the use of
...
+
+In OpenTelemetry .NET, histograms can be customized using both
+the Advice API and the View API. The actual explicit buckets used by
+the SDK are determined based on the following factors:
+
+1. The explicit buckets provided by the instrumentation owner via the Advice API.
+2. The explicit buckets defined through the View API during SDK initialization.
+3. The default buckets provided by the SDK.
+
+Here’s how the SDK determines which explicit buckets to use:
```
Comment (reyang) [inline docs/metrics/customizing-the-sdk/README.md:237]:
```suggestion
Here's how the SDK determines which explicit buckets to use:
```

**Review by reyang — APPROVED** [review]
Looks like a good step forward. I feel we might need to change the structure of the doc, the current way how explicit buckets vs. base-2 exponential buckets are organized seems to be misleading to me. Not a blocker for this PR though.
